feat(search): experimental telescope layout (closes #718)

aarnphm opened the pull request:
Definitely need some help here I spent wayy too long figuring out TS 😄 

But here is the initial look


https://github.com/jackyzha0/quartz/assets/29749331/1df60b6e-7d72-45f8-bdd9-95dec9c19026

aarnphm commented:
oh ye and also the last item and first item need to be only fetched once, and I'm figuring that out atm. So if you have any recommendation that would be great as well.

aarnphm commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -239,15 +260,43 @@ document.addEventListener("nav", async (e: unknown) => {
+      .catch((err) => {
+        console.error(err)
+      })
+    const html = p.parseFromString(contents, "text/html")
```
```suggestion
    const html = p.parseFromString(contents ?? "", "text/html")
```

tuta-amb commented:
you might want to mark this as a draft?
also, i feel like that the selection pane (on the left) is a bit too big

aarnphm committed:
feat(search): telescope-style search

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

aarnphm commented:
Here's the updated styling, fetch should work correctly now.

https://github.com/jackyzha0/quartz/assets/29749331/d7b0a933-573b-4eb2-ad93-de99d38a48ca

aarnphm committed:
chore(search): cleanup some basis and borders

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

jackyzha0 commented:
sheesh, i'm pretty packed until fri but I'll take a look as soon as I can, this looks awesome

aarnphm commented:
> sheesh, i'm pretty packed until fri but I'll take a look as soon as I can, this looks awesome

no worries take your time boss.

lucastucious commented:
How this is handled on mobile ?

aarnphm committed:
fix(search): make sure to set overflow-y

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

aarnphm commented:
> How this is handled on mobile ?

should just work on mobile afaik

benschlegel commented:
wow this is incredibly cool, very excited! may need some special treatment on mobile, as the space is a bit tight and the ride side panel seems to be blurred, but very cool!
![image](https://github.com/jackyzha0/quartz/assets/31989404/a1c8b0db-aa50-41e6-a978-1e45d4caf118)

aarnphm commented:
> wow this is incredibly cool, very excited! may need some special treatment on mobile, as the space is a bit tight and the ride side panel seems to be blurred, but very cool! 

I'm thinking only to show search item on mobile. Somewhat similar to telescope behaviour 🤔

aarnphm committed:
feat(search): shows preview on desktop only search

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

aarnphm commented:
Hmm for mobile I just disable preview. lmk what you think about this.

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +266,34 @@ document.addEventListener("nav", async (e: unknown) => {
+      .then((res) => res.text())
+      .catch((err) => {
+        console.error(err)
+      })
```
early return on error here by checking if `contents === undefined`

aarnphm replied:
This is now addressed.

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +266,34 @@ document.addEventListener("nav", async (e: unknown) => {
+    } else {
+      const previewInner = document.createElement("div")
+      previewInner.classList.add("preview-inner")
+      preview.appendChild(previewInner)
```
do we ever cleanup the created `preview-inner`? or should we replace the previous child of `preview`

aarnphm replied:
not yet I believe. The logic here somewhat similar to how popover works.

jackyzha0 replied:
popover doesn't re-fetch if the div already exists though, let's also do that here

aarnphm replied:
will do.

aarnphm replied:
Ah here it will just replace the preview-inner.

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +266,34 @@ document.addEventListener("nav", async (e: unknown) => {
     }
   }
 
+  async function fetchContent(slug: FullSlug) {
```
we should memoize results from this fn

jackyzha0 commented on quartz/components/styles/search.scss:
```diff
@@ -86,93 +87,136 @@
+              gap: 0.4rem;
+              margin: 0;
+              margin-top: 0.45rem;
+              // Offset border radius
+              margin-left: -2px;
```
should use `border-box` instead

jackyzha0 reviewed (commented):
could we gate this behind a config flag in `Search.tsx`?

aarnphm committed:
perf: add options to control layout through config

cache memoize results to avoid fetching

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

aarnphm commented:
@jackyzha0 I have updated the fetch to memorize the result. Also added flags to control this behaviour. By default it is false, which preserve the previous search behaviour.

aarnphm commented on quartz.layout.ts:
```diff
@@ -41,7 +41,7 @@ export const defaultListPageLayout: PageLayout = {
     Component.PageTitle(),
     Component.MobileOnly(Component.Spacer()),
-    Component.Search(),
+    Component.Search({ enablePreview: true }),
```
```suggestion
    Component.Search({ enablePreview: false }),
```

aarnphm commented on quartz.layout.ts:
```diff
@@ -24,7 +24,7 @@ export const defaultContentPageLayout: PageLayout = {
     Component.PageTitle(),
     Component.MobileOnly(Component.Spacer()),
-    Component.Search(),
+    Component.Search({ enablePreview: true }),
```
```suggestion
    Component.Search({ enablePreview: false }),
```

benschlegel replied:
To be consistent with all other components, we shouldn't modify the layout file here. I saw that you already set the `enablePreview: false` in `Search.tsx`, so omitting it here and leaving it as `Component.Search()` is fine.

Probably a good idea to also update the search docs to explain that this config option/feature exists as well, then people can just set it true when they want to use it

aarnphm committed:
chore: use the default configuration

aarnphm committed:
fix: correct minor type for search

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

aarnphm commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +288,42 @@ document.addEventListener("nav", async (e: unknown) => {
   }
 
+  async function fetchContent(slug: FullSlug): Promise<Element[] | undefined> {
+    if (fetchContentCache[slug]) {
```
We can do better here if we save this cache to localStorage. @jackyzha0 what's your opinion here?

benschlegel replied:
Seems a bit overkill imo, would also require you to handle localStorage not getting too full and handling cache invalidation, but let's see what Jacky says

jackyzha0 replied:
imo in memory cache is fine, let's not persist to localStorage

aarnphm committed:
fix: use datasets to query for preview

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

aarnphm commented on quartz.layout.ts:
```diff
@@ -24,7 +24,7 @@ export const defaultContentPageLayout: PageLayout = {
     Component.PageTitle(),
     Component.MobileOnly(Component.Spacer()),
-    Component.Search(),
+    Component.Search({ enablePreview: false }),
```
```suggestion
    Component.Search({ enablePreview: true }),
```

aarnphm committed:
chore: layout changes

show preview on normal layout, and only show previous layout in list page.

aarnphm committed:
fix(type): annotate search with types

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

jackyzha0 commented on quartz.layout.ts:
```diff
@@ -24,7 +24,7 @@ export const defaultContentPageLayout: PageLayout = {
     Component.PageTitle(),
     Component.MobileOnly(Component.Spacer()),
-    Component.Search(),
+    Component.Search({ enablePreview: true }),
```
```suggestion
    Component.Search(),
```

jackyzha0 commented on quartz/components/Search.tsx:
```diff
@@ -3,8 +3,18 @@ import style from "./styles/search.scss"
+}
+
+const defaultOptions: SearchOptions = {
+  enablePreview: false,
```
```suggestion
  enablePreview: true,
```

this is awesome, let's default it!!

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -71,20 +71,44 @@ function highlight(searchTerm: string, text: string, trim?: boolean) {
-document.addEventListener("nav", async (e: unknown) => {
-  const currentSlug = (e as CustomEventMap["nav"]).detail.url
+
+let fetchContentCache: Record<string, Element[] | undefined> = {}
```
```suggestion
const fetchContentCache: Map<FullSlug, Element[]> = new Map()
```

prefer `Map` for mutable

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -71,20 +71,44 @@ function highlight(searchTerm: string, text: string, trim?: boolean) {
+  if (!enablePreview) {
+    preview.style.display = "none"
+  }
+  if (searchLayout?.querySelector(`#${preview.id}`) === null) {
+    searchLayout?.appendChild(preview)
+  }
```
can we just not make `preview` and `preview-container` if `enablePreview` is false? just curious if there was a specific reason to just `display: none` instead of not adding it to the DOM

aarnphm replied:
Yes we can. I think we can chang the display to flex if enablePreview, so we don't have to figure out the display on scss.

one moment brb.

aarnphm replied:
This is now done.

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +288,42 @@ document.addEventListener("nav", async (e: unknown) => {
   }
 
+  async function fetchContent(slug: FullSlug): Promise<Element[] | undefined> {
+    if (fetchContentCache[slug]) {
```
use `Map` methods instead :)

aarnphm replied:
noted.

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +288,42 @@ document.addEventListener("nav", async (e: unknown) => {
     }
   }
 
+  async function fetchContent(slug: FullSlug): Promise<Element[] | undefined> {
```
```suggestion
  async function fetchContent(slug: FullSlug): Promise<Element[]> {
```

let's just reject instead of resolving `undefined`

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +288,42 @@ document.addEventListener("nav", async (e: unknown) => {
+        normalizeRelativeURLs(html, targetUrl)
+        return [...html.getElementsByClassName("popover-hint")]
+      })
+      .catch((err) => {
+        console.error(err)
+        throw new Error(`Could not fetch ${targetUrl}`)
+      })
```
let's not catch and let promise reject instead of returning undefined

aarnphm replied:
done.

jackyzha0 commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,6 +288,42 @@ document.addEventListener("nav", async (e: unknown) => {
+
+    fetchContentCache[slug] = contents
+    return contents
+  }
```
newline after fn def

jackyzha0 commented on quartz/components/styles/search.scss:
```diff
@@ -86,93 +86,133 @@
+            border: none !important;
           }
 
-          &:last-of-type {
-            border-bottom-left-radius: 5px;
-            border-bottom-right-radius: 5px;
-            border-bottom: 1px solid var(--lightgray);
+          & > *:not(#results-container) {
+            display: none !important;
           }
```
is there any way to do this without the `!important` selector?

aarnphm replied:
oh this is so that the preview-container doesn't show on mobile. I couldn't figure out a way not to use !important, let me test a bit and I will get back to you.

aarnphm replied:
<img width="435" alt="Screenshot 2024-01-31 at 02 49 36" src="https://github.com/jackyzha0/quartz/assets/29749331/8df81a8a-25b3-4725-b0f8-daa96a27fc9a">

Hmm this is what on phone would look like, without display: none !important

I think this looks fine on phone?

jackyzha0 replied:
how do you trigger the preview on mobile? just tap? and two taps to get to the page?

aarnphm replied:
Just tap

edit: just type I think

jackyzha0 replied:
so on mobile it only ever previews the first item? could we maybe just default to old layout on mobile

aarnphm replied:
yep sounds good.

aarnphm committed:
chore: apply jacky's suggestion

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

aarnphm commented on quartz/components/scripts/search.inline.ts:
```diff
@@ -248,10 +291,48 @@ document.addEventListener("nav", async (e: unknown) => {
 
+    if (searchLayout) {
+      // 600 here is the same value as $mobileBreakpoint in variables.scss
+      searchLayout.style.display = window.innerWidth <= 600 ? "block" : "flex"
```
@jackyzha0 is there a way to read scss variables from here? would probably best to use mobileBreakpoint instead of hardcoded here?

jackyzha0 replied:
is there any way we can just do this in CSS instead of in JS?

aarnphm replied:
we will need to use `display: none !important` 

seems like the cleanest solution.

jackyzha0 replied:
sure let's do that

aarnphm replied:
updated now.

aarnphm committed:
chore: using map API and scss

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

aarnphm committed:
fix: styling on search container view on phones

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

jackyzha0 commented on quartz.layout.ts:
```diff
@@ -41,7 +41,7 @@ export const defaultListPageLayout: PageLayout = {
     Component.PageTitle(),
     Component.MobileOnly(Component.Spacer()),
-    Component.Search(),
+    Component.Search({ enablePreview: false }),
```
only nit is to make this have no args, we should have preview on by default

```suggestion
    Component.Search(),
```

aarnphm replied:
oh right. forgot to remove this. 😄

aarnphm commented:
https://github.com/jackyzha0/quartz/assets/29749331/51666fa1-bfd8-41c0-b341-720bd294f56c

Final preview for both phone and desktop.

aarnphm committed:
Update quartz.layout.ts

Co-authored-by: Jacky Zhao <j.zhao2k19@gmail.com>

jackyzha0 commented:
few more add-ons that could be worth taking a stab at for usability
- should select/preview the first entry automatically
- mouse hover should also preview

i'll do a styling pass tomorrow

aarnphm commented:
> few more add-ons that could be worth taking a stab at for usability
> 
> * should select/preview the first entry automatically
> * mouse hover should also preview
> 
> i'll do a styling pass tomorrow

Now addressed in #774
